### DISCUSSION TITLE
Support for `HRMTextForCausalLM` architecture (HRM-Text-1B)

### DISCUSSION THREAD

**Discussion by chris-hatton** [ggml-org/llama.cpp#23415] [Ideas]
The [`sapientinc/HRM-Text-1B`](https://huggingface.co/sapientinc/HRM-Text-1B) checkpoint has been getting some attention lately - it's a 1B-parameter model that punches well above its weight on reasoning benchmarks (84.7% GSM8k, 56.5% MATH) despite being trained on only 40B tokens.

The architecture is a **dual-timescale recurrent transformer**: two standard decoder stacks (H = slow, L = fast) iterate over shared embeddings with additive state injection for `H_cycles × L_cycles` steps. In other words, the novelty is entirely in the outer recurrence loop - the inner blocks are conventional RoPE + SwiGLU + Pre-RMSNorm + gated MHA, which are already well-supported in `ggml`.

The model is Apache-2.0, already integrated in upstream `transformers`, and has no native GGUF conversion yet. Given that `llama.cpp` already handles recurrence patterns for RWKV and Mamba families, this seems like a relatively low-risk addition with a small 1B-parameter test surface.

Is there any appetite for adding this architecture? If so, what would the maintainers need to see to move from a discussion to an issue/PR - a reference GGUF conversion attempt, a logits-matching harness, or something else?

- Paper: https://sapientinc.github.io/HRM-Text/assets/HRM_Text.pdf
- Reference implementation: https://github.com/sapientinc/HRM-Text
- Model card: https://huggingface.co/sapientinc/HRM-Text-1B

**Comment by chris-hatton** [comment]
There's a claimed **-instruct** fine-tune of the base model here, that would make llama.cpp usage more interesting:
- Instruct tune: https://huggingface.co/ResulC/HRM-Text-1B-Instruct/tree/main

**Reply by Firepal** [reply]
The model creator admitted it had overfit, and is retraining it

**Comment by Firepal** [comment]
Somebody has implemented some kind of HRM-Text support independently, providing GGUFs and a patch for llama.cpp: https://huggingface.co/sinimiini/HRM-Text-1B-GGUF/

The patch doesn't seem to contain anything malicious, but I have not tried it.
Given the turnaround time, it's likely vibe-coded. I'm unfamiliar with the program's internals so I can't judge whether or not the code is any good.

**Reply by Firepal** [reply]
I built it and it seems to work. It's a 1B base model so temper your expectations.
It can CoT square roots close enough, which is neat i guess.

`./build/bin/llama-cli -m HRM-Text-1B-BF16.gguf -p "<|quad_end|><|object_ref_end|>What is the square root of 1079420?"
`

```

> <|quad_end|><|object_ref_end|>What is the square root of 1079420?

To find the square root of 1079420, we start by noting that:
- 1039² = 1079281
- 1040² = 1081600

Since 1079420 is between these two perfect squares, it lies between 1039 and 1040. To narrow down further, we use linear approximation. We approximate 1079420 as 1039² + 194. 

Using the linear approximation formula for the square root:
\[
\sqrt{1039^2 + 194} \approx 1039 + \frac{194}{2 \times 1039} = 1039 + \frac{97}{1039}
\]

Calculating \(\frac{97}{1039}\):
\[
\frac{97}{1039} \approx 0.093666
\]

Adding this to 1039:
\[
1039 + 0.093666 \approx 1039.093666
\]

To verify, we use the method of squaring and linear approximation more accurately. Let \(x = 1039.0936\). Squaring this value:
\[
(1039.0936)^2 = 1039^2 + 2 \times 1039 \times 0.0936 + (0.0936)^2
\]
\[
= 1079281 + 194.09184 + 0.0088176
\]
\[
= 1079476.0006576
\]

This is very close to 1079420, confirming the approximation. Rounding this value to four decimal places, we get approximately 1039.0936.

Thus, the square root of 1079420 is \boxed{1039.0936}.

[ Prompt: 215.8 t/s | Generation: 29.8 t/s ]
```

**Reply by Firepal** [reply]
@Resul-Caliskan You can try this patch if you're still interesting in making the instruct finetune

**Comment by x4080** [comment]
@Firepal Have you tried to train it with custom data ? How many data should it needs ? Installing now using mac m2 ..

edit:
Here's my successful step for mac
# install custom llamacpp
git clone https://github.com/ggml-org/llama.cpp
cd llama.cpp
git checkout 6a257d44633d4a752183ed778b88d2924d0a6b9d
git apply ../runtime/llama.cpp-hrm_text.patch 

# Build
cmake -B build
cmake --build build --config Release

# testrun
./build/bin/llama-cli -m ../HRM-Text-1B-GGUF/HRM-Text-1B-BF16.gguf -p "The quick brown fox" -n 32 --temp 0

# cannot use metal so -ngl 99 is making it slower, and must set -n -1 to not stop it prematurely

Maybe it should be integrated to llama cpp ? I cant find the pull request though

edit2 :
llama server works too
./build/bin/llama-server -m ../HRM-Text-1B-GGUF/HRM-Text-1B-BF16.gguf


in web ui enter prompt : <|quad_end|><|object_ref_end|>Calculate 12/3

**Reply by Firepal** [reply]
I'm on AMD so I can't train it for shit :^)
